optuna/study.py

```diff
@@ -401,7 +403,16 @@ def _create_record_and_aggregate_column(trial):
              for k in structs.FrozenTrial._ordered_fields if k in column_agg),
             [])  # type: List[Tuple[str, str]]
 
-        return pd.DataFrame(records, columns=pd.MultiIndex.from_tuples(columns))
+        df = pd.DataFrame(records, columns=pd.MultiIndex.from_tuples(columns))
+
+        if not multi_index:
+            # Flatten the `MultiIndex` columns where names are concatenated with underscores.
+            # Filtering is required to omit non-nested columns avoiding unwanted trailing
+            # underscores.
+            df.columns = [
+                '_'.join(map(lambda c: str(c), filter(lambda c: c, col))) for col in columns]
```

Comment:
`intermediate_values`'s key is `int` and this type casting is necessary.
Thank you for fixing this issue by https://github.com/optuna/optuna/commit/4583f6cc729cd02cddf6a46125c9100dad164be1.

Maybe it would be helpful for future developers if we take intermediate into account in the test case for `trials_dataframe`.

Reply:
I added intermediate values to the test. Actually, this exposed a bug that in the filtering/mapping of the columns when flattening. `intermediate_valures_0` became `intermediate_values` as the `0` was evaluated to `False` and filtered out. Fixed in https://github.com/optuna/optuna/pull/736/commits/3c94f85492fda4c772bb28e976205b2f901d4d5b#diff-4e020643693eb296936f5dd7194bd6d6R413. So, thanks a lot for the suggestion!